src/download.rs

```diff
@@ -349,9 +303,90 @@ pub fn download_to_tempfile(url: &str) -> Result<File> {
     Ok(f)
 }
 
-/// Format a size in bytes.
-fn format_bytes(count: u64) -> String {
-    Byte::from_bytes(count.into())
-        .get_appropriate_unit(true)
-        .format(1)
+struct ProgressReader<'a, R: Read> {
+    source: R,
+    length: Option<u64>,
+    artifact_type: &'a str,
+
+    position: u64,
+    last_report: Instant,
+
+    tty: bool,
+    prologue: &'static str,
+    epilogue: &'static str,
+}
+
+impl<'a, R: Read> ProgressReader<'a, R> {
+    fn new(source: R, length: Option<u64>, artifact_type: &'a str) -> Result<Self> {
+        let tty = isatty(stderr().as_raw_fd()).chain_err(|| "checking if stderr is a TTY")?;
+        Ok(ProgressReader {
+            source,
+            length,
+            artifact_type,
+
+            position: 0,
+            last_report: Instant::now(),
+
+            tty,
+            // If stderr is a tty, draw a status line that updates itself in
+            // place.  The prologue leaves a place for the cursor to rest
+            // between updates.  The epilogue writes three spaces to cover
+            // the switch from e.g.  1000 KiB to 1 MiB, and then uses CR to
+            // return to the start of the line.
+            //
+            // Otherwise, stderr is being read by another process, e.g.
+            // journald, and fanciness may confuse it.  Just log regular
+            // lines.
+            prologue: if tty { "> " } else { "" },
+            epilogue: if tty { "   \r" } else { "\n" },
+        })
+    }
+
+    /// Format a size in bytes.
+    fn format_bytes(count: u64) -> String {
+        Byte::from_bytes(count.into())
+            .get_appropriate_unit(true)
+            .format(1)
+    }
+}
+
+impl<'a, R: Read> Read for ProgressReader<'a, R> {
+    fn read(&mut self, buf: &mut [u8]) -> result::Result<usize, io::Error> {
+        let count = self.source.read(buf)?;
+        self.position += count as u64;
+        if self.last_report.elapsed() >= Duration::from_secs(1)
+            || self.length == Some(self.position)
+        {
+            self.last_report = Instant::now();
+            match self.length {
+                Some(length) if length > 0 => eprint!(
+                    "{}Read {} {}/{} ({}%){}",
+                    self.prologue,
+                    self.artifact_type,
+                    Self::format_bytes(self.position),
+                    Self::format_bytes(length),
```

Comment:
I think you can compute this cosmetic string when building the reader and directly cache it there, to avoid recomputing on each iteration.